File: backend/src/main/java/com/ll/hotel/domain/hotel/option/service/HotelOptionService.java
package com.ll.hotel.domain.hotel.option.service;

import com.ll.hotel.domain.hotel.option.dto.request.OptionRequest;
import com.ll.hotel.domain.hotel.option.entity.HotelOption;
import com.ll.hotel.domain.hotel.option.repository.HotelOptionRepository;
import com.ll.hotel.global.exceptions.ErrorCode;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@RequiredArgsConstructor
public class HotelOptionService {
    private final HotelOptionRepository hotelOptionRepository;

    @Transactional
    public HotelOption add(OptionRequest optionRequest) {
        HotelOption hotelOption = HotelOption
                .builder()
                .name(optionRequest.name())
                .build();

        return hotelOptionRepository.save(hotelOption);
    }

    public List<HotelOption> findAll() {
        return hotelOptionRepository.findAll();
    }

    public HotelOption findById(Long id) {
        return hotelOptionRepository.findById(id)
                .orElseThrow(ErrorCode.HOTEL_OPTION_NOT_FOUND::throwServiceException);
    }

    @Transactional
    public void modify(HotelOption hotelOption, OptionRequest optionRequest) {
        hotelOption.setName(optionRequest.name());
    }

    public void flush() {
        hotelOptionRepository.flush();
    }

    public void delete(HotelOption hotelOption) {

        if (!hotelOption.getHotels().isEmpty()) {
            ErrorCode.OPTION_IN_USE.throwServiceException();
        }

        hotelOptionRepository.delete(hotelOption);
    }
}

File: backend/src/main/java/com/ll/hotel/domain/hotel/option/service/HotelOptionService.kt
package com.ll.hotel.domain.hotel.option.service

import com.ll.hotel.domain.hotel.option.dto.request.OptionRequest
import com.ll.hotel.domain.hotel.option.dto.response.OptionResponse
import com.ll.hotel.domain.hotel.option.entity.HotelOption
import com.ll.hotel.domain.hotel.option.repository.HotelOptionRepository
import com.ll.hotel.global.exceptions.ErrorCode
import jakarta.transaction.Transactional
import org.springframework.stereotype.Service

@Service
class HotelOptionService(
    private val hotelOptionRepository: HotelOptionRepository
) {
    @Transactional
    fun add(optionRequest: OptionRequest): OptionResponse {
        val hotelOption: HotelOption = hotelOptionRepository.save(HotelOption(optionRequest.name))
        return OptionResponse.from(hotelOption)
    }

    fun findAll(): List<OptionResponse> =
        hotelOptionRepository.findAll().map { OptionResponse.from(it) }

    fun findById(id: Long): HotelOption = hotelOptionRepository.findById(id)
        .orElseThrow(ErrorCode.HOTEL_OPTION_NOT_FOUND::throwServiceException)

    @Transactional
    fun modify(id: Long, optionRequest: OptionRequest): OptionResponse {
        val hotelOption = findById(id)
        hotelOption.name = optionRequest.name
        hotelOptionRepository.flush()
        return OptionResponse.from(hotelOption)
    }

    fun delete(id: Long) {
        val hotelOption = findById(id)

        if (hotelOption.hotels.isNotEmpty()) {
            ErrorCode.OPTION_IN_USE.throwServiceException()
        }
        hotelOptionRepository.delete(hotelOption)
    }
}